sync/syncer.go

```diff
@@ -145,97 +148,175 @@ func (s *Syncer) Synchronise() {
+
+	//fetch layer hash from each peer
+	m, err := s.fetchLayerHashes(currenSyncLayer)
+
+	if err != nil {
+		s.Error("could not get LayerHashes for layer: %v", currenSyncLayer)
+		return nil, err
+	}
+
+	//fetch ids for each hash
+	blockIds, err := s.fetchLayerBlockIds(m, currenSyncLayer)
 	if err != nil {
 		s.Error("could not get layer block ids %v", currenSyncLayer, err)
 		return nil, err
 	}
-	output := FetchBlocks(s, blockIds)
+
+	blocksArr, err := s.fetchFullBlocks(blockIds)
+	if err != nil {
+		s.Error("could not get layer blocks %v", currenSyncLayer, err)
+		return nil, err
+	}
+
+	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
+}
+
+func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
+	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		mb := out.(*types.MiniBlock)
-		foundTxs, missing := s.GetTransactions(mb.TxIds)
-		fetchedTxs := s.fetchTxs(missing)
-		txs := make([]*types.SerializableTransaction, len(mb.TxIds))
-		for _, t := range mb.TxIds {
-			if tx, ok := foundTxs[t]; ok {
-				txs = append(txs, tx)
-			} else {
-				txs = append(txs, fetchedTxs[t])
-			}
+
+		//sync Transactions
+		txs, err := s.Txs(mb)
+		if err != nil {
+			s.Warning(fmt.Sprintf("failed fetching block %v transactions ", err))
+			continue
+		}
+
+		//sync ATxs
+		atxs, err := s.ATXs(mb)
+		if err != nil {
+			s.Warning(fmt.Sprintf("failed fetching block %v activation transactions ", err))
+			continue
 		}
 
-		block := &types.Block{BlockHeader: mb.BlockHeader, Txs: txs, ATXs: mb.ATXs}
-		s.Debug("add block to layer %v", block)
-		s.AddBlock(block)
+		block := &types.Block{BlockHeader: mb.BlockHeader, Txs: txs, ATXs: atxs}
+		eligible, err := s.BlockEligible(&block.BlockHeader)
+		if err != nil {
+			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
+			continue
+		}
+		if !eligible {
+			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
+			continue
+		}
+		s.Info("add block to layer %v", block)
+		if err := s.AddBlock(block); err != nil {
+			s.Warning(fmt.Sprintf("could not add %v", block.ID()), err)
+			continue
+		}
+		s.Info("added block to layer %v", block)
 		blocksArr = append(blocksArr, block)
 
 	}
 
-	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
-}
-
-type peerHashPair struct {
-	peer p2p.Peer
-	hash []byte
+	return blocksArr, nil
 }
 
-func (s *Syncer) getLayerBlockIDs(index types.LayerID) (chan types.BlockID, error) {
-
-	m, err := s.fetchLayerHashes(index)
+func (s *Syncer) Txs(mb *types.MiniBlock) ([]*types.SerializableTransaction, error) {
+	foundTxs, missing := s.GetTransactions(mb.TxIds)
+	//map and sort txs
+	txMap := make(map[types.TransactionId]*types.SerializableTransaction)
+	if len(missing) > 0 {
+		for out := range s.fetchWithFactory(TxReqFactory(missing), 1) {
```

Comment:
this does not validate that all missing txs we're actually received

Reply:
not yet

Reply:
It's validated below, isn't it?